drgn_tools/list_lru.py

```diff
@@ -0,0 +1,261 @@
+        else:
+            # not lru.memcg_aware
+            if has_member(lru, "ext"):
+                li = lru.ext.node[nid].lru.list
+            else:
+                li = lru.node[nid].lru.list
+            for entry in list_for_each_entry(type, li.address_of_(), member):
+                yield entry
+
+
+def list_lru_kmem_to_memcgidx(prog: Program, kvm: IntegerLike) -> IntegerLike:
+    """
+    Return the memcg index of the list_lru entry.
+    Return -1 if the list_lru is not memcg enabled or value could not be found.
+    Memory cgroups for slab allocation are per object. This code expects a slab
+    allocated kvm and the MEMCG_DATA_KMEM case is NOT covered in this routine.
+
+    :param prog: Kernel being debugged
+    :param kvm: address of a list_lru
+    :return: memcg index, -1 means not found
+    """
+    page = virt_to_page(prog, kvm)
+    cpage = compound_head(page)
+    # page_objcgs_check() MEMCG_DATA_OBJCGS memcg are managed per object
+    if has_member(cpage, "memcg_data") or has_member(cpage, "obj_cgroups"):
+        if has_member(cpage, "memcg_data"):
+            memcg_data = cpage.memcg_data
+        else:
+            # cast to an integer for the MEMCG_DATA_KMEM test.
+            memcg_data = cast("unsigned long", cpage.obj_cgroups)
+        if memcg_data & MEMCG_DATA_OBJCGS:
+            objcgrp = Object(
+                prog, "struct obj_cgroup **", memcg_data - MEMCG_DATA_OBJCGS
+            )
+            # offset of object calculation
+            pvm = page_to_virt(cpage)
+            kvm = Object(prog, "void *", kvm)
+            if has_member(cpage, "slab_cache"):
+                slab_cache = cpage.slab_cache
+            else:
+                # v5.17 (uek8) moved the kmem_cache to a new slab structure.
+                # and since v6.10 the slab pages are identified by a page type
+                if PageSlab(cpage):
+                    slab = Object(prog, "struct slab *", cpage)
+                    slab_cache = slab.slab_cache
+                else:
+                    return -1
+            objoffset = (kvm - pvm) / slab_cache.size
+            memcgrp = objcgrp[objoffset].memcg
+            if memcgrp == NULL(prog, "struct mem_cgroup *"):
+                return -1
+            else:
+                return memcgrp.kmemcg_id
+        else:
+            return -1
+    else:
+        # Before v5.13
+        scache = cpage.slab_cache
+        if scache == NULL(prog, "struct kmem_cache *"):
+            return -1
+        else:
+            return cpage.slab_cache.memcg_params.memcg.kmemcg_id
+
+
+def list_lru_kmem_to_nodeid(prog: Program, kvm: IntegerLike) -> IntegerLike:
+    """
+    Return the NUMA node id of the list_lru entry.
+
+    :param prog: Kernel being debugged
+    :param kvm: address of a list_lru entry
+    :return: NUMA node id
+    """
+    page = virt_to_page(prog, kvm)
+    cpage = compound_head(page)
+    #
+    # Stephen Brennan routine to convert a page to node
```

Comment:
no need to include my name here. I hope I'll actually be able to come up with a better way, so I don't even necessarily want my name associated with this code 😜